crates/fluvio-spu/src/replication/leader/replica_state.rs

```diff
@@ -311,11 +320,16 @@ where
     /// write records to storage
     /// then update our follower's leo
     #[instrument(skip(self, records, notifiers))]
-    pub async fn write_record_set<R: BatchRecords>(
+    pub async fn write_record_set(
         &self,
-        records: &mut RecordSet<R>,
+        records: &mut RecordSet<RawRecords>,
         notifiers: &FollowerNotifier,
     ) -> Result<(Offset, Offset, usize)> {
+        self.transform(records).await?;
```

Comment:
Could you tell me why transformation is performed here instead of `produce_handler`? Replication logic should be decoupled event processing.  Meaning all write goto leader

Reply:
- `produce_handler` handles only input from producers, but potentially, it could be some topic to topic data movements (derived topic or something similar) which also should uphold deduplication guarantees.
 - `produce_handler` doesn't have access to a particular replica's lifecycle (whether it's added or promoted or removed), so keeping track of all SM contexts in there will be hard and error-prone.
 - in general, we plan to use this replica sm context for schema management as well, there could be more. From the architectural point of view, it will be too much overloading for  `produce_handler` which essentially is a service, which main responsibility is to handle API requests.

Reply:
This is per design.  SPU's `produce` lifecycle shouldn't deal with the replica's lifecycle.  Our replication model is a single master writer.  The replication should not deal with SM, which is only relevant for a writer.  Otherwise, this will result in inconsistent replication.  The promotion/demotion replication cycle should not impact write processing.

In the future, replication needs to move to its own crate and needs to be firewalled from the rest of SPU similar to storage.

Reply:
> The replication should not deal with SM

This is a question to the entire feature design then, then to the code. In the design https://github.com/infinyon/roadmap/blob/main/rfd/deduplication-rfd.md it was clearly stated that `LeaderReplicaState` owns the SM context.

> Otherwise, this will result in inconsistent replication

The replication model is not affected at all. The SM runs only on LeaderReplicaState. The transformation happens on write before data is sent to the storage. All replication events read from the storage.

> The promotion/demotion replication cycle should not impact write processing.

It doesn't. Why do you think it does?

Reply:
After offline discussion, this approach seems to be good enough for now since we don't have good leader replica abstraction which is necessary for https://github.com/infinyon/fluvio/issues/3235.